src/librustc_trait_selection/traits/error_reporting/suggestions.rs

```diff
@@ -316,6 +327,100 @@ fn suggest_restriction(
 }
 
 impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
+    /// Try to find possible types that would satisfy the bounds in the type param to give an
+    /// appropriate turbofish suggestion.
+    fn get_turbofish_suggestions(
+        &self,
+        obligation: &PredicateObligation<'tcx>,
+        data: ty::TraitPredicate<'tcx>,
+        self_ty: Ty<'tcx>,
+    ) -> Vec<String> {
+        let mut turbofish_suggestions = FxHashSet::default();
+        self.tcx.for_each_relevant_impl(data.trait_ref.def_id, self_ty, |impl_def_id| {
+            let param_env = ty::ParamEnv::empty();
+            let param_env = param_env.subst(self.tcx, data.trait_ref.substs);
+            let ty = self.next_ty_var(TypeVariableOrigin {
+                kind: TypeVariableOriginKind::NormalizeProjectionType,
+                span: DUMMY_SP,
+            });
+
+            let impl_substs = self.fresh_substs_for_item(obligation.cause.span, impl_def_id);
+            let trait_ref = self.tcx.impl_trait_ref(impl_def_id).unwrap();
+            let trait_ref = trait_ref.subst(self.tcx, impl_substs);
+
+            // Require the type the impl is implemented on to match
+            // our type, and ignore the impl if there was a mismatch.
+            let cause = traits::ObligationCause::dummy();
+            let eq_result = self.at(&cause, param_env).eq(trait_ref.self_ty(), ty);
```

Comment:
So I was going to say that `.eq` and other side-effects need to be wrapped in `self.probe(|_| {...})` (which will roll back the side-effects, although you have to be careful to not keep anything with e.g. `Ty` in it from inside the `probe`, outside).

But it looks like you're equating an inference variable with `trait_ref.self_ty()`? That doesn't do anything useful, this will always succeed and the effect will be that `ty` will now always resolve to whatever `trait_ref.self_ty()` is.

I was thinking you're equating with the original `self_ty`, which would make sense, but you're not.

Overall I don't think this code does anything, all the filtering is provided by `for_each_relevant_impl` above (you should try it on something with blanket impls or maybe `impl<T> ... for &T` because neither is filtered by it).